chalk-solve/src/clauses/builtin_traits/unsize.rs

```diff
@@ -0,0 +1,449 @@
+            let auto_trait_ids_b: Vec<_> = auto_trait_ids(db, bounds_b).collect();
+
+            let may_apply = principal_a == principal_b
+                && auto_trait_ids_b
+                    .iter()
+                    .all(|id_b| auto_trait_ids_a.iter().any(|id_a| id_a == id_b));
+
+            if !may_apply {
+                return;
+            }
+
+            // COMMENT FROM RUSTC:
+            // Require that the traits involved in this upcast are **equal**;
+            // only the **lifetime bound** is changed.
+            //
+            // FIXME: This condition is arguably too strong -- it would
+            // suffice for the source trait to be a *subtype* of the target
+            // trait. In particular, changing from something like
+            // `for<'a, 'b> Foo<'a, 'b>` to `for<'a> Foo<'a, 'a>` should be
+            // permitted.
+            // <...>
+            // I've modified this to `.eq` because I want to continue rejecting
+            // that [`old-lub-glb-object.rs`] test (as we have
+            // done for quite some time) before we are firmly comfortable
+            // with what our behavior should be there. -nikomatsakis
+            // ------------------
+
+            // Filter out auto traits of source that are not present in target
+            // and change source lifetime to target lifetime
+            //
+            // This new type should be equal to target type.
+            let source_ty = TyData::Dyn(DynTy {
+                bounds: bounds_a.map_ref(|bounds| {
+                    QuantifiedWhereClauses::from(
+                        interner,
+                        bounds.iter(interner).filter(|bound| {
+                            let trait_id = match bound.trait_id() {
+                                Some(id) => id,
+                                None => return true,
+                            };
+
+                            if auto_trait_ids_a.iter().all(|&id_a| id_a != trait_id) {
+                                return true;
+                            }
+                            auto_trait_ids_b.iter().any(|&id_b| id_b == trait_id)
+                        }),
+                    )
+                }),
+                lifetime: lifetime_b.clone(),
+            })
+            .intern(interner);
+
+            // Check that new source is equal to target
+            let eq_goal = EqGoal {
+                a: source_ty.cast(interner),
+                b: target_ty.clone().cast(interner),
+            }
+            .cast(interner);
+
+            // FIXME(areredify) change this to outlives once #419 lands
+            let lifetime_outlives_goal = EqGoal {
+                a: lifetime_a.clone().cast(interner),
+                b: lifetime_b.clone().cast(interner),
+            }
+            .cast(interner);
+
+            builder.push_clause(trait_ref.clone(), [eq_goal, lifetime_outlives_goal].iter());
```

Comment:
This seems...weird. I think.

Imagine that our original source and target are `dyn Trait<P0, P1> + AutoX + 'a` and `dyn Trait<P2, P3> + 'b`, respectively. So, what we've essentially done is taken the source made a "new" source `dyn Trait<P0, P1> + 'b` and here we generate a clause `(dyn Trait<P0, P1> + AutoX + 'a): Unsize<(dyn Trait<P2, P3> + 'b)> :- (dyn Trait<P0, P1> + 'b) = (dyn Trait<P2, P3> + 'b), 'a == 'b`. Why don't we just generate the clause `(dyn Trait<P0, P1> + AutoX + 'a): Unsize<(dyn Trait<P2, P3> + 'b)> :- P0 == P2, P1 ==P3, 'a == 'b`?

Reply:
The other thing I guess I've kind of left out is other `WhereClause`s. Maybe that's what this is for?

Reply:
Because it might be a a higher-order bound, if you have 
```
dyn for<'a, 'b> Fn(&'a u32, &'b u32) -> dyn for<'a> Fn(&'a u32, &'a u32)
```
you can't generate the clause you are talking about